oci/archive/reader.go

```diff
@@ -0,0 +1,116 @@
+package archive
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/image/v5/internal/tmpdir"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+// Reader keeps the temp directory the oci archive will be untarred to and the manifest of the images
+type Reader struct {
+	manifest      *imgspecv1.Index
+	tempDirectory string
+	path          string // The original, user-specified path; not the maintained temporary file, if any
+}
+
+// NewReader creates the temp directory that keeps the untarred archive from src.
+// The caller should call .Close() on the returned object.
+func NewReader(ctx context.Context, sys *types.SystemContext, src string) (*Reader, error) {
+	// TODO: This can take quite some time, and should ideally be cancellable using a context.Context.
+	arch, err := os.Open(src)
+	if err != nil {
+		return nil, err
+	}
+	defer arch.Close()
+
+	dst, err := ioutil.TempDir(tmpdir.TemporaryDirectoryForBigFiles(sys), "oci")
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp directory")
+	}
+
+	reader := Reader{
+		tempDirectory: dst,
+		path:          src,
+	}
+
+	succeeded := false
+	defer func() {
+		if !succeeded {
+			reader.Close()
+		}
+	}()
+	if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+		return nil, errors.Wrapf(err, "error untarring file %q", dst)
+	}
+
+	indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer indexJSON.Close()
+	reader.manifest = &imgspecv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+		return nil, err
+	}
+	succeeded = true
+	return &reader, nil
+}
```

Comment:
`docker/archive.Reader` needs to provide a `NewReaderForReference` for libimage, AFAICS to turn an user-supplied `docker-archive:…@0` into a tag used in c/storage on `podman load`.

Will something like that be necessary here? (Cc: @vrothberg )

Reply:
I update NewReader to take in a reference, if that is wrong I can break them apart tot two different functions `NewReader` and NewReaderForReference`.